crates/resolver-tests/src/lib.rs

```diff
@@ -227,51 +247,217 @@ fn sat_at_most_one(solver: &mut impl varisat::ExtendFormula, vars: &[varisat::Va
 ) -> HashMap<K, Vec<varisat::Var>> {
-    // no two packages with the same links set
+    // no two packages with the same keys set
     let mut by_keys: HashMap<K, Vec<varisat::Var>> = HashMap::new();
     for (p, v) in data {
         by_keys.entry(p).or_default().push(v)
     }
     for key in by_keys.values() {
-        sat_at_most_one(cnf, key);
+        sat_at_most_one(solver, key);
     }
     by_keys
 }
 
+fn find_compatible_dep_summaries_by_name_in_toml(
+    pkg: &Summary,
+    by_name: &HashMap<InternedString, Vec<Summary>>,
+) -> IndexMap<InternedString, Vec<Summary>> {
+    let empty_vec = vec![];
+
+    pkg.dependencies()
+        .iter()
+        .map(|dep| {
+            let name_in_toml = dep.name_in_toml();
+
+            let compatible_summaries = by_name
+                .get(&dep.package_name())
+                .unwrap_or(&empty_vec)
+                .iter()
+                .filter(|s| dep.matches_id(s.package_id()))
+                .filter(|s| dep.features().iter().all(|f| s.features().contains_key(f)))
+                .cloned()
+                .collect::<Vec<_>>();
+
+            (name_in_toml, compatible_summaries)
+        })
+        .collect()
+}
+
+fn process_feature_value(
+    solver: &mut varisat::Solver<'_>,
+    var_for_is_packages_used: &HashMap<PackageId, varisat::Var>,
+    var_for_is_packages_features_used: &HashMap<PackageId, HashMap<InternedString, varisat::Var>>,
+    feature_value: &FeatureValue,
+    bound_var: varisat::Var,
+    pkg_feature_var_map: &HashMap<InternedString, varisat::Var>,
+    compatible_dep_summaries_by_name_in_toml: &IndexMap<InternedString, Vec<Summary>>,
+) {
+    match *feature_value {
+        FeatureValue::Feature(other_feature_name) => {
+            solver.add_clause(&[
+                bound_var.negative(),
+                pkg_feature_var_map[&other_feature_name].positive(),
+            ]);
+        }
+        FeatureValue::Dep { dep_name } => {
+            let dep_clause = compatible_dep_summaries_by_name_in_toml[&dep_name]
+                .iter()
+                .map(|dep| var_for_is_packages_used[&dep.package_id()].positive())
+                .chain([bound_var.negative()])
+                .collect::<Vec<_>>();
+
+            solver.add_clause(&dep_clause);
+        }
+        FeatureValue::DepFeature {
+            dep_name,
+            dep_feature: dep_feature_name,
+            weak,
+        } => {
+            for dep in &compatible_dep_summaries_by_name_in_toml[&dep_name] {
+                let dep_var = var_for_is_packages_used[&dep.package_id()];
+                let dep_feature_var =
+                    var_for_is_packages_features_used[&dep.package_id()][&dep_feature_name];
+
+                solver.add_clause(&[
+                    bound_var.negative(),
+                    dep_var.negative(),
+                    dep_feature_var.positive(),
+                ]);
+            }
+
+            if !weak {
```

Comment:
This looks basically correct. I wish I could make a stronger statement. I wouldn't be surprised if there are corner cases about situations where there are features and dependencies by the same name or around distinguishing between the default feature and a feature called "default". But I don't know how to find these corner cases except comparing cargo and that SAT resolver across a large number of examples. Which is not part of this PR because it's follow-up work.

Some of that follow-up work will look like:
- making the Prop Tests output features that might hit this corner case.
- I found a lot of corner cases the hard way when comparing cargo and the PubGrub based resolver. The minimize examples are at [Eh2406/pubgrub-crates-benchmark/index_ron](https://github.com/Eh2406/pubgrub-crates-benchmark/tree/main/out/index_ron). We could turn those test examples into hand coded tests in `crates/resolver-tests/tests/resolve.rs`.
- We could bring some of that approach to this project and have a command to run `resolve_and_validated` for each version of every crate on crates.io.

Anyway just some context and potential follow-up work.

Reply:
> where there are features and dependencies by the same name or around distinguishing between the default feature and a feature called "default"

This is managed when constructing the `Summary`:

https://github.com/rust-lang/cargo/blob/01e1ab55dd7f391cef809c05a50030ae3bb018d8/src/cargo/core/summary.rs#L189

An implicit feature `dep1 = ["dep:dep1"]` from an optional dependency is only created if the feature `dep1` is not already defined, and if the feature `default` is not already defined, an implicit feature `default = []` is added.
Since the SAT resolver receives `Summary` objects, theses corner cases don't need to be handled.